### PR TITLE
scoped tokens

### PR DESCRIPTION
**Pull request by deads2k**
This demonstrates how scopes could be attached to a user token and then filtered by an authorizer that wraps our existing authorizer.  It creates: 
 1. `special:whoami` - call users/~ to prove identity
 2. `special:access-check` - call self-SARs to check access of the user (not the token)
 3. `clusterrole:<clusterrole name>:<namespace to allow the cluster role, * means all>` - allows access based on cluster role rules, but only within the specified namespace.  (I might make that a comma-delimited list.)

If we like the general approach, I can finish off the list:
 - [x] unit tests
 - [x] integration tests that create scoped tokens

And do the followups:
 - [x] make SARs scope aware https://trello.com/c/o0pWuPmd/698-5-auth-scopes-scoped-access-checks-sars-tsars
 - [x] make TSARs for checking a token https://trello.com/c/o0pWuPmd/698-5-auth-scopes-scoped-access-checks-sars-tsars
 - [ ] add the other two scope types (namespace-wide and cluster-wide explicit)

 - [x] tighten clusterrole/role name validation to the set of characters allowed in scope names? may not be possible... failing validation on one role would prevent the entire policy from being persisted  Won't do this, but the scope validation prevents people from using them in a scope.  If you want a scope for it, rename your role.
 - [x] expand scope evaluator interface to provide human descriptions for scopes (used in grant flow) https://github.com/openshift/origin/pull/8807
 - [x] expand scope evaluator interface to provide validation (used when requesting scopes or validating a user-created token) https://github.com/openshift/origin/pull/8805
 - [ ] prevent escalating resources from being included by default:
remove *, secrets, oauthtokens, oauthauthorizationtokens, etc
 - [ ] decide how a scope that grants a role including escalating resources would be specified

@pweil- @liggitt @smarterclayton 
@jwforres @jimmidyson @stefwalter you guys knew people who were looking to get openshift API tokens to perform actions for a user, right?

### CODE REVIEW

**Comment by smarterclayton** [comment]
Aren't there scope names defined in oauth that we have to support as well?
Do we need aliases for them?

**Comment by smarterclayton** [comment]
Also, Peter the Great is not amused.

**Comment by liggitt** [comment]
There are standard scopes in OIDC (https://openid.net/specs/openid-connect-basic-1_0.html#Scopes)

**Comment by deads2k** [comment]
I'll need to reread the openid bits.  I thought those were used by a client to request certain claims be provided.  The user was still asked, but the token held the information, not any lasting access.  If the access does have to be provided, chances are we need new endpoints to match the spec format, right?

File: pkg/authorization/authorizer/scope/authorizer.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func NewAuthorizer(delegate defaultauthorizer.Authorizer, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) defaultauthorizer.Authorizer {
+	return &scopeAuthorizer{delegate: delegate, clusterPolicyGetter: clusterPolicyGetter}
+}
+
+func (a *scopeAuthorizer) Authorize(ctx kapi.Context, passedAttributes defaultauthorizer.AuthorizationAttributes) (bool, string, error) {
+	user, exists := kapi.UserFrom(ctx)
+	if !exists {
+		return false, "", fmt.Errorf("no user")
```
Comment (stefwalter) [inline pkg/authorization/authorizer/scope/authorizer.go]: Suggest a more descriptive error message here if there's any risk of it ending up at an API caller.

**Comment by stefwalter** [comment]
@deads2k The main use case that registry users needed is a long lived token that they can plug into 'docker login'. That is, not have to run a docker login command with a new token every 24 hours.

Would we need the registry container need to be modified to do a special access check?

CC @aweiteka I believe this is about the long lived tokens you wanted earlier.

**Comment by deads2k** [comment]
> Would we need the registry container need to be modified to do a special access check?

This would allow a user to provide a token that is only good for manipulating images instead of trusting the registry with all their access, including things like modifying group membership as an example.  This is also something we'd want for a user created token flow, but doesn't provide one in and of itself.

**Comment by deads2k** [comment]
> There are standard scopes in OIDC (https://openid.net/specs/openid-connect-basic-1_0.html#Scopes)

Yeah, we don't need to add those yet.  This structure doesn't preclude them from working, but for now we aren't close to being an openID provider.  If we want to be, that should be a separate objective.

**Comment by aweiteka** [comment]
> This would allow a user to provide a token that is only good for manipulating images

The docker push/pull API commands are ultimately what I'm looking for supporting so this subset of permissions or access control sounds like it could satisfy that.

**Comment by deads2k** [comment]
@openshift/api-review  this is on my next list.  Do we like the shape of these?  It's basically, `<category>:<category-specific-format>`, which initial support for curated scopes and cluster-role scopes applied to a namespace.

**Comment by smarterclayton** [comment]
If we supported OIDC in the future - those would be category-less?

File: pkg/authorization/authorizer/scope/converter.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+func clusterRoleEvaluator(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error) {
+	ctx := kapi.WithNamespace(kapi.NewContext(), namespace)
+	policy, err := clusterPolicyGetter.GetClusterPolicy(ctx, "default")
+	if err != nil {
+		return nil, err
+	}
+
+	tokens := strings.SplitN(scope, ":", 2)
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("bad format for %v", scope)
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: `"bad format for scope %v"` so I can find it when I have to debug it

File: pkg/authorization/authorizer/scope/converter.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	// namespaces can't have colons, but roles can.  reverse, split on first, reverse back.
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: This is a bit weird, why not use LastIndexOf?

File: pkg/authorization/authorizer/scope/converter.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+var scopeEvaluators = map[string]scopeEvaluator{
+	SpecialIndicator:     specialEvaluator,
+	ClusterRoleIndicator: clusterRoleEvaluator,
+}
+
+// scopes are in the format
+// <indicator><indicator choice>
+// we have the following formats:
+// clusterwide:<comma-delimitted verbs>:<comma-delimitted resources>
+// namespace:<namespace name>:<comma-delimitted verbs>:<comma-delimitted resources>
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: Needs to be updated?  Also, `comma-delimited`

File: pkg/authorization/authorizer/scope/authorizer.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	scopes, ok := uncastScopes.([]string)
+	if !ok || len(scopes) == 0 {
+		return a.delegate.Authorize(ctx, passedAttributes)
+	}
+
+	nonFatalErrors := []error{}
+
+	namespace, _ := kapi.NamespaceFrom(ctx)
+	// scopeResolutionErrors aren't fatal.  If any of the scopes we find allow this, then the overall scope limits allow it
+	rules, err := ScopesToRules(scopes, namespace, a.clusterPolicyGetter)
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/authorizer.go]: Is this going to be expensive to calculate repeatedly?

File: pkg/authorization/authorizer/scope/converter.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+// scopeEvaluators map prefixes to a function that handles that prefix
+var scopeEvaluators = map[string]scopeEvaluator{
+	SpecialIndicator:     specialEvaluator,
+	ClusterRoleIndicator: clusterRoleEvaluator,
+}
+
+// scopes are in the format
+// <indicator><indicator choice>
+// we have the following formats:
+// clusterwide:<comma-delimitted verbs>:<comma-delimitted resources>
```
Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: `cluster-scope:` and `namespace-scope`?

Comment (smarterclayton) [inline pkg/authorization/authorizer/scope/converter.go]: I'm not in love with clusterwide - `cluster` would be ok with me as well.

File: pkg/authorization/authorizer/scope/authorizer.go (hunk 5 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/authorizer.go]: > Is this going to be expensive to calculate repeatedly?

Not moreso than authorization rules and checked less frequently.

File: pkg/authorization/authorizer/scope/converter.go (hunk 4 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: > Needs to be updated? Also, comma-delimited

More of a TODO.  You said these were secondary objectives for kind of crazy people.

**Comment by deads2k** [comment]
comments addressed, upstream pick update, tests written.

This is ready for review.

@liggitt you or a delegating victim?

**Comment by deads2k** [comment]
[test]

**Comment by brianwcook** [comment]
what happens if you login using a long lived token when you have openid connect provider enabled?  
You get a token and then the provider shuts off your account.  Is this accounted for?

**Comment by deads2k** [comment]
> what happens if you login using a long lived token when you have openid connect provider enabled?

The cluster-admin determines the maximum token lifespan, so they determine long-livedness. Right now, there is no re-authentication check required for the lifespan of the token.  I think its 24 hours by default.

This pull doesn't make the problem any better or worse.

**Comment by deads2k** [comment]
@smarterclayton this blocks oauth exposure.  How's the API?

**Comment by smarterclayton** [comment]
I've seen worse.  Approved.

**Comment by deads2k** [comment]
> I've seen worse. Approved.

I feel like I should find the movie review translating tool they use for commercials. :)

@pweil- got a reviewer in mind?

**Comment by pweil-** [comment]
@liggitt and @sgallagher PTAL and give a :+1: or :-1:

File: pkg/authorization/authorizer/scope/converter.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+var discoveryRules = []authorizationapi.PolicyRule{
+	{
+		Verbs: sets.NewString("get"),
+		NonResourceURLs: sets.NewString(
+			// API discovery/negotiation
+			"/api", "/api/*",
+			"/apis", "/apis/*",
+			"/oapi", "/oapi/*",
```
Comment (soltysh) [inline pkg/authorization/authorizer/scope/converter.go]: Can we make constants out of it, there's a TODO in [master.go](https://github.com/openshift/origin/blob/2a4f294f506743d18bec3d3db26cfe23d16a42cb/pkg/cmd/server/origin/master.go#L111-L114) to have that prefix being configurable.

File: pkg/authorization/authorizer/scope/converter_test.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+	"fmt"
+	"strings"
+	"testing"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+)
+
+func TestSpecialEvaluator(t *testing.T) {
```
Comment (soltysh) [inline pkg/authorization/authorizer/scope/converter_test.go]: This test is very fragile, since it's testing only number of rules you get from the evaluator. There are no checks on the output PolicyRules being produced by those. The same concern applies to cluster evaluator.

**Comment by soltysh** [comment]
@deads2k with yours and @liggitt explanation yesterday during planning and some time with your code I was able to grasp the idea. Thanks :beer:

File: pkg/authorization/authorizer/scope/converter_test.go (hunk 8 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter_test.go]: > This test is very fragile, since it's testing only number of rules you get from the evaluator. There are no checks on the output PolicyRules being produced by those. The same concern applies to cluster evaluator.

The cure may be worse than the disease.  You either duplicate them (bad idea) or have separate methods for getting them (which is pretty ugly).  The integration test checks which permissions are actually selected.

Comment (soltysh) [inline pkg/authorization/authorizer/scope/converter_test.go]: The integration tests addresses only one special token, in that case, what about others and clusterrole one?

Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter_test.go]: > The integration tests addresses only one special token, in that case, what about others and clusterrole one?

Do you gain value by having me type it twice?  If not, do you gain value by having me duplicate the switch-case?  If not, then how would I structure the test?

Comment (soltysh) [inline pkg/authorization/authorizer/scope/converter_test.go]: I'm more worried about not having these parts covered with tests and if somebody will try to change one might break it accidentally. But, I won't block the PR for that.

File: pkg/authorization/authorizer/scope/converter.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		),
+	},
+}
+
+const (
+	SpecialIndicator       = "special:"
+	ClusterRoleIndicator   = "clusterrole:"
+	ClusterWideIndicator   = "clusterwide:"
+	NamespaceWideIndicator = "namespace:"
```
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: need to make sure we limit ourselves to valid characters for scopes
* no quote, no backslash, no space
* colon, alphanumeric characters, comma is ok (full list of allowed chars is %x21 / %x23-5B / %x5D-7E)

File: pkg/authorization/authorizer/scope/converter.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+// scopeEvaluator takes a scope and returns the rules that express it
+type scopeEvaluator func(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error)
+
+// scopeEvaluators map prefixes to a function that handles that prefix
+var scopeEvaluators = map[string]scopeEvaluator{
+	SpecialIndicator:     specialEvaluator,
+	ClusterRoleIndicator: clusterRoleEvaluator,
```
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: rather than a prefix:evaluator map, I'd prefer a list of evaluators, and make one of the return values whether the evaluator recognized/evaluated the scope. I'd like to avoid the "special:" prefix if possible

Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: would be good to expand the evaluator func to an interface that can also return a human description of the scope

Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: > rather than a prefix:evaluator map, I'd prefer a list of evaluators, and make one of the return values whether the evaluator recognized/evaluated the scope, and short-circuit the list when one recognizes the scope format. I'd like to avoid the "special:" prefix if possible

I've regretted not having a map in the past (think our resource to handler maps in some admission plugins), but I don't feel strongly about that.

I do really want a "special" list for curated options instead of making roles to satisfy them.  Having a curated list for the standard or most common options is easier to describe than a role which is mutable.  In addition, while roles may seem nice, I see enumerated permissions as more reliable since roles are at the mercy of the cluster-admin and may not even exist in all clusters.  I think that roles are what people think they want, but in terms of portability they will not be particularly reliable for contained units.

Short version, "special" exists to provide a guaranteed way of saying "exactly this set of permissions, no matter how this cluster is configured" and proof of identity and can I do this are both important enough to warrant inclusion.

Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: I don't disagree with a specia list, I'd just make them unnamespaced :)

Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: > I don't disagree with a special list, I'd just make them unnamespaced :)

I've regretted that in the past too because it compresses the overall namespace.  Given a display name and description in the API (that seems very reasonable), do you still care? No regular user will see these.

**Comment by deads2k** [comment]
So if I make a constant and eliminate the "special:" prefix (I still think we're likely to regret that some day.  Unscoped things always cause problems eventually.), I'll get a merge?

**Comment by liggitt** [comment]
I'm ok with the approach. possible name adjustments:

specific scopes:
```
user:info
user:check-access
```
parameterized scopes:
```
role:<cluster role>:<effective namespace>
role:<cluster role>:*
```

eventually:
```
user:create-token
user:request-project
local-role:<role>:<namespace>
```

followups:
* tighten clusterrole/role name validation to the set of characters allowed in scope names? may not be possible... failing validation on one role would prevent the entire policy from being persisted
* expand scope evaluator interface to provide human descriptions for scopes
* prevent escalating resources from being included by default:
  * remove `*`, `secrets`, `oauthtokens`, `oauthauthorizationtokens`, etc
* decide how a scope that grants a role including escalating resources would be specified

File: pkg/authorization/authorizer/scope/converter.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+			{Verbs: sets.NewString("create"), Resources: sets.NewString("subjectaccessreviews", "localsubjectaccessreviews"), AttributeRestrictions: &authorizationapi.IsPersonalSubjectAccessReview{}},
+		}, nil
+	default:
+		return nil, fmt.Errorf("unrecognized scope: %v", scope)
+	}
+}
+
+// clusterrole:<clusterrole name>:<namespace to allow the cluster role, * means all>
+func clusterRoleEvaluator(scope, namespace string, clusterPolicyGetter rulevalidation.ClusterPolicyGetter) ([]authorizationapi.PolicyRule, error) {
+	tokens := strings.SplitN(scope, ":", 2)
```
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: nit, use LastIndex instead of splitting and reshuffling

File: pkg/authorization/authorizer/scope/authorizer.go (hunk 5 above)
Comment (liggitt) [inline pkg/authorization/authorizer/scope/authorizer.go]: checking attributes against the rules as we build them and short-circuiting seems like a good future optimization

**Comment by deads2k** [comment]
> I'm ok with the approach. possible name adjustments:

Ok, those look fine.

@sgallagher all comments so far are minor and I'll take care of them tomorrow.  Do you have anything structural?

File: pkg/authorization/authorizer/scope/converter.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+		return nil, fmt.Errorf("bad format for scope %v", scope)
+	}
+
+	// namespaces can't have colons, but roles can.  pick last.
+	lastColonIndex := strings.LastIndex(tokens[1], ":")
+	if lastColonIndex <= 0 || lastColonIndex == (len(tokens[1])-1) {
+		return nil, fmt.Errorf("bad format for scope %v", scope)
+	}
+	roleName := tokens[1][0:lastColonIndex]
+	scopeNamespace := tokens[1][lastColonIndex+1:]
```
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter.go]: make sure len(roleName) and len(scopeNamespace) are > 0? Otherwise, empty namespace matches cluster-scoped resource requests

File: pkg/authorization/authorizer/scope/converter_test.go [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+			numRules:  2,
+		},
+		{
+			name:            "getter-error",
+			policyGetterErr: fmt.Errorf("some bad thing happened"),
+			namespace:       "current-ns",
+			scopes:          []string{ClusterRoleIndicator + "admin:two:current-ns"},
+			err:             `some bad thing happened`,
+			numRules:        1,
+		},
```
Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter_test.go]: other edge cases:
* empty namespace segment
* empty role segment
* valid role containing a colon

Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter_test.go]: Because an empty role might be a valid name?  Unlikely.

File: pkg/authorization/authorizer/scope/converter.go (hunk 12 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: Empty role name fails lookup, empty namespace fails last index check above.

File: pkg/authorization/authorizer/scope/converter_test.go (hunk 13 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter_test.go]: Both of those are in the list under bad format 3 and 4.

Comment (liggitt) [inline pkg/authorization/authorizer/scope/converter_test.go]: Ah, missed the colon embedded in the constant

File: pkg/authorization/authorizer/scope/converter.go (hunk 7 above)
Comment (deads2k) [inline pkg/authorization/authorizer/scope/converter.go]: > Can we make constants out of it, there's a TODO in master.go to have that prefix being configurable.

done

**Comment by deads2k** [comment]
comments address [merge]
